KAFKA-8725: Improve LogCleanerManager#grabFilthiestLog error handling

stanislavkozlovski opened the pull request:
KAFKA-7215 improved the log cleaner error handling to mitigate thread death but missed one case. Exceptions in `grabFilthiestCompactedLog` still cause the thread to die. 

This patch improves handling to ensure that errors in that function still mark a partition as uncleanable and do not crash the thread.

stanislavkozlovski committed:
KAFKA-8725: Improve LogCleanerManager#grabFilthiestLog error handling

KAFKA-7215 improved the log cleaner error handling to mitigate thread death but missed one case. Exceptions in `grabFilthiestCompactedLog` still cause the thread to die.

This patch improves handling to ensure that errors in that function still mark a partition as uncleanable and do not crash the thread.

stanislavkozlovski commented:
JDK 11/Scala 2.12 failed with one `kafka.api.SaslOAuthBearerSslEndToEndAuthorizationTest.testNoDescribeProduceOrConsumeWithoutTopicDescribeAcl` - surely unrelated

stanislavkozlovski commented:
cc @hachikuji @mumrah - this is related to https://issues.apache.org/jira/browse/KAFKA-8724

hachikuji commented on core/src/main/scala/kafka/log/LogCleanerManager.scala:
```diff
@@ -39,6 +39,9 @@ private[log] case object LogCleaningInProgress extends LogCleaningState
 private[log] case object LogCleaningAborted extends LogCleaningState
 private[log] case class LogCleaningPaused(pausedCount: Int) extends LogCleaningState
 
+private[log] case class ErroneousLogException(log: Log,
```
How about `LogCleaningException`?

stanislavkozlovski replied:
Naming things is hard

hachikuji commented on core/src/main/scala/kafka/log/LogCleaner.scala:
```diff
@@ -342,10 +342,15 @@ class LogCleaner(initialConfig: CleanerConfig,
+            case e: ErroneousLogException =>
+              e.log
+            case _ =>
+              if (currentLog.isEmpty) {
```
It seems like the only reason we have `currentLog` is for this catch block. I'm wondering if we can remove it and rely on the `ErroneousLogException` more generally. In other words, rather than setting `currentLog` before calling `cleanLog`, we can just let it raise `ErroneousLogException` in the case of an error. What do you think?

stanislavkozlovski replied:
Sounds good to me. The new logic polluted the method with try/catch blocks, so I split it into two methods

Please double-check I didn't introduce a bug with this refactor, as it seems easy to slip through

hachikuji commented on core/src/test/scala/unit/kafka/log/LogCleanerManagerTest.scala:
```diff
@@ -61,6 +61,62 @@ class LogCleanerManagerTest extends Logging {
+      cleanerManager.grabFilthiestCompactedLog(time).get
+      fail(s"Expected grabFilthiestCompactedLog call to throw an $ErroneousLogException")
+    } catch {
+      case e: ErroneousLogException =>
```
nit: use `assertThrows`

stanislavkozlovski replied:
I would like to assert the `assertEquals(log, e.log)` check though, is there a way to achieve that with assertThrows?

hachikuji replied:
`assertThrows` returns the caught exception.

stanislavkozlovski replied:
Thanks, I didn't know that. Will use it

hachikuji commented on core/src/test/scala/unit/kafka/log/LogCleanerManagerTest.scala:
```diff
@@ -61,6 +61,62 @@ class LogCleanerManagerTest extends Logging {
+      logDirFailureChannel = new LogDirFailureChannel(10))
+    fillUpLog(log2, log2SegmentsCount)
+
+    val tp3 = new TopicPartition("A", 2)
```
Can this test be simplified at all? Do we need three partitions?

hachikuji commented on core/src/test/scala/unit/kafka/log/LogCleanerManagerTest.scala:
```diff
@@ -511,6 +560,23 @@ class LogCleanerManagerTest extends Logging {
+    LogConfig(logProps)
+  }
+
+  private def fillUpLog(log: Log, segmentsCount: Int): Unit = {
```
nit: maybe `fillLogSegments`?

hachikuji reviewed (commented):
Thanks, this makes sense. Left a few comments.

stanislavkozlovski commented:
Thanks for the review. LMK how this looks

stanislavkozlovski commented on core/src/main/scala/kafka/log/LogCleaner.scala:
```diff
@@ -304,53 +304,58 @@ class LogCleaner(initialConfig: CleanerConfig,
+        cleanFilthiestLog()
+      } catch {
+        case e @ (_: ThreadShutdownException | _: ControlThrowable) => throw e
+        case e: LogCleaningException =>
```
Note that any unexpected exceptions will still crash the cleaner thread. This PR isn't changing the functionality there

hachikuji commented on core/src/main/scala/kafka/log/LogCleanerManager.scala:
```diff
@@ -39,6 +39,9 @@ private[log] case object LogCleaningInProgress extends LogCleaningState
 
+private[log] case class LogCleaningException(log: Log,
+                                             private val message: String,
+                                             private val cause: Throwable) extends Exception(message, cause)
```
May not matter too much, but does it make sense to extend `KafkaException`?

hachikuji reviewed (commented):
Thanks for the fixes. Left a minor comment. And note I responded to your question about `assertThrows`.

stanislavkozlovski commented:
I had to rebase with trunk due to conflicts and decided to squish all commits into one for ease of rebasing

stanislavkozlovski commented:
retest this please

hachikuji commented:
retest this please

stanislavkozlovski commented:
Rebased with trunk

stanislavkozlovski commented:
All 5 kafka.log.LogCleanerParameterizedIntegrationTest.cleanerConfigUpdateTest tests fail with an error similar to:
```
java.lang.AssertionError: log cleaner should have processed up to offset 286, but lastCleaned=0
```
I'll give it a better look later and come back

stanislavkozlovski commented:
I found the error, I had removed the catch block handling of `case e @ (_: ThreadShutdownException | _: ControlThrowable)` and erroneously re-throwed it as a LogCleaningException which would get handled and not shut down the thread:
https://github.com/apache/kafka/pull/7127/files#diff-d7330411812d23e8a34889bee42fedfeL343

Because I had deleted my fork, I have re-opened a new PR with these changes - https://github.com/apache/kafka/pull/7475.
